app/models/exercise_map.rb

```diff
@@ -0,0 +1,51 @@
+class ExerciseMap
+  def initialize(user, track)
+    @user = user
+    @track = track
+  end
+
+  def core_exercises
+    map.select(&:core?)
+  end
+
+  private
+  attr_reader :user, :track
+
+  def map
+    @map ||= exercises.map do |e|
+      ExerciseMapNode.new(
+        exercise: e,
+        unlocks: exercises_to_unlock_for(e)
+      )
+    end
+  end
+
+  def exercises
+    @exercises ||= track.
+      exercises.
+      map { |e| ExerciseWithSolution.new(e, solution_for(e)) }
+  end
+
+  def exercises_to_unlock_for(exercise)
+    if exercise.auto_approve?
+      return exercises.
+        select(&:side?).
+        select { |e| e.unlocked_by.blank? }[0, 10]
```

Comment:
@kntsoriano I think this needs to be this **as well** as the bit below. An auto-approve might unlock exercises as well as the bonus ones.